polkadot/node/core/prospective-parachains/src/fragment_chain/mod.rs

```diff
@@ -0,0 +1,1047 @@
+	fn from(value: &'a messages::HypotheticalCandidate) -> Self {
+		match value {
+			messages::HypotheticalCandidate::Complete {
+				receipt,
+				persisted_validation_data,
+				..
+			} => Self::Complete {
+				receipt: Cow::Borrowed(receipt),
+				persisted_validation_data: Cow::Borrowed(persisted_validation_data),
+			},
+			messages::HypotheticalCandidate::Incomplete {
+				parent_head_data_hash,
+				candidate_relay_parent,
+				..
+			} => Self::Incomplete {
+				relay_parent: *candidate_relay_parent,
+				parent_head_data_hash: *parent_head_data_hash,
+			},
+		}
+	}
+}
+
+pub struct FragmentNode {
+	fragment: Fragment<'static>,
+	pub candidate_hash: CandidateHash,
+	cumulative_modifications: ConstraintModifications,
+}
+
+impl FragmentNode {
+	fn relay_parent(&self) -> Hash {
+		self.fragment.relay_parent().hash
+	}
+}
+
+/// This is a chain of candidates based on some underlying storage of candidates and a scope.
+///
+/// All nodes in the chain must be either pending availability or within the scope. Within the scope
+/// means it's built off of the relay-parent or an ancestor.
+pub(crate) struct FragmentChain {
+	scope: Scope,
+
+	pub chain: Vec<FragmentNode>,
+
+	candidates: HashSet<CandidateHash>,
+
+	// Index from head data hash to candidate hashes with that head data as a parent.
+	by_parent_head: HashMap<Hash, CandidateHash>,
+	// Index from head data hash to candidate hashes outputting that head data.
+	by_output_head: HashMap<Hash, CandidateHash>,
+}
+
+impl FragmentChain {
+	/// Create a new [`FragmentChain`] with given scope and populated from the storage.
+	pub fn populate(scope: Scope, storage: &CandidateStorage) -> Self {
+		gum::trace!(
+			target: LOG_TARGET,
+			relay_parent = ?scope.relay_parent.hash,
+			relay_parent_num = scope.relay_parent.number,
+			para_id = ?scope.para,
+			ancestors = scope.ancestors.len(),
+			"Instantiating Fragment Chain",
+		);
+
+		let mut fragment_chain = Self {
+			scope,
+			chain: Vec::new(),
+			candidates: HashSet::new(),
+			by_parent_head: HashMap::new(),
+			by_output_head: HashMap::new(),
+		};
+
+		fragment_chain.populate_chain(storage);
+
+		fragment_chain
+	}
+
+	/// Get the scope of the Fragment Chain.
+	pub fn scope(&self) -> &Scope {
+		&self.scope
+	}
+
+	/// Returns an O(n) iterator over the hashes of candidates contained in the
+	/// chain.
+	pub(crate) fn candidates(&self) -> impl Iterator<Item = CandidateHash> + '_ {
+		self.candidates.iter().cloned()
+	}
+
+	/// Returns the number of candidates in the chain
+	pub(crate) fn len(&self) -> usize {
+		self.candidates.len()
+	}
+
+	/// Whether the candidate exists.
+	pub(crate) fn contains_candidate(&self, candidate: &CandidateHash) -> bool {
+		self.candidates.contains(candidate)
+	}
+
+	/// Try accumulating more candidates onto the chain.
+	///
+	/// Candidates can only be added if they build on the already existing chain.
+	pub(crate) fn repopulate(&mut self, storage: &CandidateStorage) {
```

Comment:
This public method name seems wrong, I was assuming we remove everthing than add everthing back in. All that it seems to do is extend the chain based on new stuff in storage. 

`extend_from_storage` should do more justice.

Reply:
agreed 👍🏻